File: api/check-election.js
const fetch = require('node-fetch');
const xml2js = require('xml2js');
const fs = require('fs');
const path = require('path');

module.exports = async (req, res) => {
try {
const feedUrl = path.join(__dirname, 'test.xml');
const xml = fs.readFileSync(feedUrl, 'utf-8');
const keywords = [
'pope elected', 'new pope', 'white smoke', 'new pontiff',
'habemus papam', 'papal conclave', 'vatican elects',
'successor to francis', 'new vatican leader',
'conclave selects pope', 'st. peters square',
'new head of catholic church'
];

php
Zkopírovat
Upravit
const response = await fetch(feedUrl);
const xml = await response.text();
const parser = new xml2js.Parser();
const feed = await parser.parseStringPromise(xml);

const items = feed.rss.channel[0].item || [];
const now = new Date();
const recentThreshold = 1000 * 60 * 60 * 48; // posledních 48 hodin

const matches = items.filter(item => {
  const pubDate = new Date(item.pubDate[0]);
  const age = now - pubDate;
  if (age > recentThreshold) return false;

  const title = item.title?.[0]?.toLowerCase() || '';
  const desc = item.description?.[0]?.toLowerCase() || '';
  return keywords.some(kw => title.includes(kw) || desc.includes(kw));
});

const responsePayload = {
  popeElected: matches.length > 0,
  articles: matches.map(item => ({
    title: item.title[0],
    link: item.link[0],
    pubDate: item.pubDate[0]
  }))
};

res.setHeader('Content-Type', 'application/json');
res.status(200).json(responsePayload);
} catch (err) {
console.error('Election check error:', err);
res.status(500).json({ error: 'Election check failed' });
}
};

File: api/check-election.go
package main

import (
"context"
"encoding/json"
"encoding/xml"
"net/http"
"strings"
"time"
)

// --- RSS structs ---
type RSS struct {
Channel Channel xml:"channel"
}
type Channel struct {
Items []Item xml:"item"
}
type Item struct {
Title string xml:"title"
Link string xml:"link"
Description string xml:"description"
PubDate string xml:"pubDate"
}

type Article struct {
Title string json:"title"
Link string json:"link"
PubDate string json:"pubDate"
}

type Result struct {
PopeElected bool json:"popeElected"
Articles []Article json:"articles"
}

// --- Vercel handler ---
func Handler(w http.ResponseWriter, r *http.Request) {
url := "https://search.cnbc.com/rs/search/combinedcms/view.xml?partnerId=wrss01&id=100003114"
resp, err := http.Get(url)
if err != nil {
http.Error(w, "Failed to fetch RSS", http.StatusInternalServerError)
return
}
defer resp.Body.Close()

go
Zkopírovat
Upravit
var rss RSS
if err := xml.NewDecoder(resp.Body).Decode(&rss); err != nil {
	http.Error(w, "Failed to parse RSS", http.StatusInternalServerError)
	return
}

keywords := []string{
	"pope elected", "new pope", "white smoke", "new pontiff",
	"habemus papam", "papal conclave", "vatican elects",
	"successor to francis", "new vatican leader",
	"conclave selects pope", "st. peter's square",
	"new head of catholic church",
}

twoDaysAgo := time.Now().Add(-48 * time.Hour)
var matches []Article

for _, item := range rss.Channel.Items {
	pubTime, _ := parseTime(item.PubDate)
	if pubTime.Before(twoDaysAgo) {
		continue
	}
	title := strings.ToLower(item.Title)
	desc := strings.ToLower(item.Description)
	for _, kw := range keywords {
		if strings.Contains(title, kw) || strings.Contains(desc, kw) {
			matches = append(matches, Article{
				Title:   item.Title,
				Link:    item.Link,
				PubDate: item.PubDate,
			})
			break
		}
	}
}

result := Result{
	PopeElected: len(matches) > 0,
	Articles:    matches,
}

w.Header().Set("Content-Type", "application/json")
json.NewEncoder(w).Encode(result)
}

// helper
func parseTime(dateStr string) (time.Time, error) {
t, err := time.Parse(time.RFC1123Z, dateStr)
if err == nil {
return t, nil
}
return time.Parse(time.RFC1123, dateStr)
}